### PR TITLE
Building on Windows

### PR DESCRIPTION
**Pull request by lissyx**

### CODE REVIEW

File: native_client/args.h [hunk 1]
```diff
@@ -1,7 +1,11 @@
 #ifndef __ARGS_H__
 #define __ARGS_H__
 
+#if defined(_MSC_VER)
+#include "getopt_win.h"
```
Comment (lissyx) [inline native_client/args.h]: needs to be named differenly otherwise some tensorflow build will pick-it when building `generate_trie` and others

File: native_client/client.cc [hunk 2]
```diff
@@ -261,20 +266,25 @@ main(int argc, char **argv)
     }
   }
 
+#ifndef NO_SOX
```
Comment (lissyx) [inline native_client/client.cc]: Looks like this is one we forgot when doing Android

File: native_client/javascript/Makefile [hunk 3]
```diff
@@ -32,7 +36,7 @@ configure: deepspeech_wrap.cxx package.json
 	$(NODE_BUILD_TOOL) configure $(NODE_BUILD_VERBOSE)
 
 build: configure deepspeech_wrap.cxx
-	AS=$(AS) CC=$(CC) CXX=$(CXX) LD=$(LD) CFLAGS="$(CFLAGS)" CXXFLAGS="$(CXXFLAGS)" LDFLAGS="$(RPATH_NODEJS) $(LDFLAGS)" LIBS="$(LIBS)" $(NODE_BUILD_TOOL) $(NODE_PLATFORM_TARGET) $(NODE_ABI_TARGET) rebuild $(NODE_BUILD_VERBOSE)
+	AS=$(AS) CC=$(CC) CXX=$(CXX) LD=$(LD) CFLAGS="$(CFLAGS)" CXXFLAGS="$(CXXFLAGS)" LDFLAGS="$(RPATH_NODEJS) $(LDFLAGS)" LIBS=$(LIBS) $(NODE_BUILD_TOOL) $(NODE_PLATFORM_TARGET) $(NODE_ABI_TARGET) rebuild $(NODE_BUILD_VERBOSE)
```
Comment (lissyx) [inline native_client/javascript/Makefile]: The quotes around it makes MSVC add an extra space when picking the lib name, and this break the build

File: native_client/python/Makefile [hunk 4]
```diff
@@ -8,7 +8,7 @@ bindings-clean:
 
 bindings-build:
 	pip install --quiet $(PYTHON_PACKAGES) wheel==0.31.0 setuptools==39.1.0
-	AS=$(AS) CC=$(CC) CXX=$(CXX) LD=$(LD) CFLAGS="$(CFLAGS)" LDFLAGS="$(LDFLAGS_NEEDED) $(RPATH_PYTHON)" MODEL_LDFLAGS="$(LDFLAGS_DIRS)" MODEL_LIBS="$(LIBS)" $(PYTHON_PATH) $(NUMPY_INCLUDE) python ./setup.py build_ext $(PYTHON_PLATFORM_NAME)
+	PATH=$(TOOLCHAIN):$$PATH AS=$(AS) CC=$(CC) CXX=$(CXX) LD=$(LD) CFLAGS="$(CFLAGS)" LDFLAGS="$(LDFLAGS_NEEDED) $(RPATH_PYTHON)" MODEL_LDFLAGS="$(LDFLAGS_DIRS)" MODEL_LIBS="$(LIBS)" $(PYTHON_PATH) $(NUMPY_INCLUDE) python ./setup.py build_ext $(PYTHON_PLATFORM_NAME)
```
Comment (lissyx) [inline native_client/python/Makefile]: Required to make sure that `cl.exe` is being picked and that `link.exe` is also picked from the MSVC toolchain and not from the `msys2` binaries

File: native_client/python/setup.py [hunk 5]
```diff
@@ -39,11 +39,27 @@ class BuildExtFirst(build):
...
+
+    if os.name == 'nt':
+        return a.split('.lib')[0:1]
+
 ds_ext = Extension('deepspeech._impl',
          ['impl.i'],
          include_dirs = [ numpy_include, '../' ],
-         library_dirs = list(map(lambda x: x.strip(), os.getenv('MODEL_LDFLAGS', '').split('-L')[1:])),
-         libraries = list(map(lambda x: x.strip(), os.getenv('MODEL_LIBS', '').split('-l')[1:])))
+         library_dirs = list(map(lambda x: x.strip(), lib_dirs_split(os.getenv('MODEL_LDFLAGS', '')))),
```
Comment (lissyx) [inline native_client/python/setup.py]: We were splitting on GCC/clang's way to pass variables, we need different behavior with MSVC linker.

File: native_client/python/setup.py [hunk 6]
```diff
@@ -39,11 +39,27 @@ class BuildExtFirst(build):
...
+
+    if os.name == 'nt':
+        return []
+
+def libs_split(a):
+    if os.name == 'posix':
+        return a.split('-l')[1:]
+
+    if os.name == 'nt':
+        return a.split('.lib')[0:1]
```
Comment (lissyx) [inline native_client/python/setup.py]: Somehow, Python build system force-add an extra `.lib`, I could not find a good explanation why.

File: taskcluster/test-win-opt-base.tyml [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        export TASKCLUSTER_TMP_DIR="$TASKCLUSTER_TASK_DIR/tmp" &&
+        export PIP_DEFAULT_TIMEOUT=60 &&
+        (mkdir $TASKCLUSTER_TASK_DIR || rm -fr $TASKCLUSTER_TASK_DIR/*) && cd $TASKCLUSTER_TASK_DIR &&
+        env &&
+        ln -s $USERPROFILE/msys64 $TASKCLUSTER_TASK_DIR/msys64 &&
+        git clone --quiet ${event.head.repo.url} $TASKCLUSTER_TASK_DIR/DeepSpeech/ds/ &&
+        cd $TASKCLUSTER_TASK_DIR/DeepSpeech/ds && git checkout --quiet ${event.head.sha} &&
+        cd $TASKCLUSTER_TASK_DIR &&
+        pacman --noconfirm -R bsdtar &&
+        pacman --noconfirm -S tar vim &&
```
Comment (lissyx) [inline taskcluster/test-win-opt-base.tyml]: We need `vim` to get `xxd` when checking the output of tests.

File: taskcluster/win-build.sh [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    BAZEL_BUILD_FLAGS="${BAZEL_OPT_FLAGS} ${BAZEL_EXTRA_FLAGS}"
+    BAZEL_ENV_FLAGS="TF_NEED_CUDA=0"
+fi
+
+SYSTEM_TARGET=host-win
+
+do_bazel_build
+
+if [ "${cuda}" = "--cuda" ]; then
+    cp ${DS_ROOT_TASK}/DeepSpeech/tf/bazel-bin/native_client/liblibdeepspeech.so.ifso ${DS_ROOT_TASK}/DeepSpeech/tf/bazel-bin/native_client/libdeepspeech.so.if.lib
```
Comment (lissyx) [inline taskcluster/win-build.sh]: No idea as to why the lib name gets broken like that, but the symbols inside were fine

File: taskcluster/win-opt-base.tyml [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+      TENSORFLOW_BUILD_ARTIFACT: ${build.tensorflow}
+      DEEPSPEECH_TEST_MODEL: https://queue.taskcluster.net/v1/task/${training}/artifacts/public/output_graph.pb
+
+  command:
+    - >-
+      "C:\Program Files\7-zip\7z.exe" x -txz -so msys2-base-x86_64.tar.xz |
+      "C:\Program Files\7-zip\7z.exe" x -o%USERPROFILE% -ttar -aoa -si
+    - .\msys64\usr\bin\bash.exe --login -cx "exit"
+    - .\msys64\usr\bin\bash.exe --login -cx "pacman --noconfirm -Syu"
+    - echo .\msys64\usr\bin\bash.exe --login -cxe "
```
Comment (lissyx) [inline taskcluster/win-opt-base.tyml]: Check the `echo`, it's a magic trick.

File: taskcluster/win-opt-base.tyml [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+      cd $TASKCLUSTER_TASK_DIR/DeepSpeech/ds && git checkout --quiet ${event.head.sha} &&
+      ln -s $TASKCLUSTER_TASK_DIR/DeepSpeech/ds/native_client/ $TASKCLUSTER_TASK_DIR/DeepSpeech/tf/native_client &&
+      cd $TASKCLUSTER_TASK_DIR &&
+      pacman --noconfirm -R bsdtar &&
+      pacman --noconfirm -S tar make &&
+      $TASKCLUSTER_TASK_DIR/DeepSpeech/ds/${build.scripts.build} &&
+      $TASKCLUSTER_TASK_DIR/DeepSpeech/ds/${build.scripts.package} ;
+      echo \"export TASKCLUSTER_TASK_EXIT_CODE=$?\" > $USERPROFILE/tc-exit.sh &&
+      cd $TASKCLUSTER_TASK_DIR/../ && rm -fr tc-workdir/ && exit $TASKCLUSTER_TASK_EXIT_CODE" | cmd
+      /k ""C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat"" x64
```
Comment (lissyx) [inline taskcluster/win-opt-base.tyml]: The trick is here: to properly perform some build tasks, MSVC requires `vcvarsall.bat` to be ran to setup some env variables like `LIB`, `INCLUDE`, etc. Sourcing that from bash is not working, so we need to setup the env before calling bash. Thats what the `echo "" | cmd /k ...` does here, it will run the script first and keep cmd running, and the echo piping will run our own script that starts a `bash -c`.

The double quotes are the documented way to pass `cmd.exe` path with spaces.

File: taskcluster/win-opt-base.tyml [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+    - .\msys64\usr\bin\bash.exe --login -cxe "source $USERPROFILE/tc-exit.sh &&
```
Comment (lissyx) [inline taskcluster/win-opt-base.tyml]: Because of the trick above, we were not able to properly `exit $TASKCLUSTER_EXIT_CODE`, this would get mangled into `cmd.exe` so the easiest is to output it into a shell in the user's profile (which is different for each scheduled task, so no overlapping), and return it this way. It helps avoiding silent failures.

File: tc-tests-utils.sh [hunk 12]
```diff
@@ -226,7 +251,19 @@ run_tflite_basic_inference_tests()
   assert_correct_ldc93s1 "${phrase_pbmodel_nolm}"
 }
 
-run_all_inference_tests()
+run_netframework_inference_tests()
```
Comment (lissyx) [inline tc-tests-utils.sh]: The output contains much more informations, so we just check that what we want is here.

File: tc-tests-utils.sh [hunk 13]
```diff
@@ -299,14 +341,53 @@ generic_download_tarxz()
...
+{
+  PROJECT_NAME=$1
+  if [ -z "${PROJECT_NAME}" ]; then
+    exit "Please call with a valid PROJECT_NAME"
+    exit 1
+  fi;
+
+  nuget="${PROJECT_NAME}.${DS_VERSION}.nupkg"
+
+  export PATH=$PATH:$(cygpath ${ChocolateyInstall})/bin
```
Comment (lissyx) [inline tc-tests-utils.sh]: required for `nuget` binary

File: tc-tests-utils.sh [hunk 14]
```diff
@@ -299,14 +341,53 @@ generic_download_tarxz()
...
+
+  mkdir -p "${TASKCLUSTER_TMP_DIR}/repo/"
+  mkdir -p "${TASKCLUSTER_TMP_DIR}/ds/"
+
+  wget -O - "${DEEPSPEECH_ARTIFACTS_ROOT}/${nuget}" | gunzip > "${TASKCLUSTER_TMP_DIR}/${PROJECT_NAME}.${DS_VERSION}.nupkg"
```
Comment (lissyx) [inline tc-tests-utils.sh]: `gunzip` because TaskCluster infra does `gzip` and it looks like `wget` does not respect / deflate automagically

File: native_client/client.cc [hunk 15]
```diff
@@ -150,29 +161,25 @@ GetAudioBuffer(const char* path)
...
-#ifdef __ANDROID__
+#ifdef NO_SOX
   // FIXME: Hack and support only 16kHz mono 16-bits PCM
   FILE* wave = fopen(path, "r");
 
   size_t rv;
 
   unsigned short audio_format;
   fseek(wave, 20, SEEK_SET); rv = fread(&audio_format, 2, 1, wave);
-  assert(rv == 2);
```
Comment (reuben) [inline native_client/client.cc]: Why did you remove all the asserts?

File: tc-tests-utils.sh [hunk 16]
```diff
@@ -390,6 +471,8 @@ is_patched_bazel()
 {
   bazel_version=$(bazel version | grep 'Build label:' | cut -d':' -f2)
 
+  bazel shutdown
```
Comment (reuben) [inline tc-tests-utils.sh]: Any reason why you're not using `shutdown_bazel` here?

File: native_client/client.cc (hunk 15 above)
Comment (lissyx) [reply native_client/client.cc]: They fail on Windows while we get proper values being read. We have other assets at the end, and it's only for specific usecase, I considered it was not worth spending more time than needed

Comment (reuben) [reply native_client/client.cc]: OK

File: tc-tests-utils.sh (hunk 16 above)
Comment (lissyx) [reply tc-tests-utils.sh]: Yes, `shutdown_bazel` ensures using the bazel output directory, if we issue a `bazel shutdown` with a wrong path (i.e., it should be symetric to the `bazel` call that started the process) then it will leave dangling processes.
